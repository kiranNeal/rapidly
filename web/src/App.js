import React, { createContext, useReducer } from 'react';

import './App.css';
import Header from './components/Header';

const AppContext = createContext(null);

function App() {
  return (
    <div className="App">
      <Header />
    </div>
  );
}

export { App, AppContext };
