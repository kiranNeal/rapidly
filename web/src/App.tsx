import React, { createContext, useReducer, Dispatch } from 'react';
import reducer, { initialState } from './states/application/applicationReducer';
import {
  ApplicationState,
  Action
} from './states/application/applicationTypes';
import Header from './components/Header';
import RequestOptions from './components/RequestOptions';

interface InitContextProps {
  state: ApplicationState;
  dispatch: Dispatch<Action>;
}

const AppContext = createContext({} as InitContextProps);

const App: React.FC = () => {
  const [state, dispatch] = useReducer(reducer, initialState);
  return (
    <AppContext.Provider value={{ state, dispatch }}>
      <div className="App">
        <Header />
        <RequestOptions></RequestOptions>
      </div>
    </AppContext.Provider>
  );
};

export { App, AppContext };
