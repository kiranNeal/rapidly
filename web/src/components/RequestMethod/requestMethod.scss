#request_method__dropdown {
    width: 120px;
}