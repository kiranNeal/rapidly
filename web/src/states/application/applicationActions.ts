import {
  actionTypes,
  SetMethodAction,
  ResetStateAction
} from './applicationTypes';

export function setMethod(method: string): SetMethodAction {
  return {
    type: actionTypes.setMethod,
    payload: method
  };
}

export function resetState(): ResetStateAction {
  return {
    type: actionTypes.resetState
  };
}
