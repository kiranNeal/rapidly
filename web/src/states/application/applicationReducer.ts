import {
  Action,
  actionTypes,
  ApplicationState,
  ResetStateAction,
  SetMethodAction
} from './applicationTypes';

export const initialState = {
  url: '',
  method: 'get'
};

// The Type Guard Functions
function isResetStateAction(action: Action): action is ResetStateAction {
  return action.type === actionTypes.resetState;
}
function isSetMethodAction(action: Action): action is SetMethodAction {
  return action.type === actionTypes.setMethod;
}

export default function reducer(
  state: ApplicationState = initialState,
  action: Action
): ApplicationState {
  if (isSetMethodAction(action)) {
    return {
      ...state,
      method: action.payload
    };
  }
  if (isResetStateAction(action)) {
    return { ...initialState };
  }

  return state;
}
