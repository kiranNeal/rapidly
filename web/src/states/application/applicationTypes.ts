export const actionTypes = {
  resetState: 'RESET_METHOD',
  setMethod: 'SET_METHOD'
};

interface MethodTypes {
  [key: string]: string;
}

export const methodTypes: MethodTypes = {
  get: 'GET',
  post: 'POST',
  delete: 'DELETE',
  put: 'PUT',
  patch: 'PATCH'
};

export interface ApplicationState {
  url: string;
  method: typeof methodTypes.get;
}

export interface Action {
  type: string;
}

export interface ResetStateAction extends Action {
  type: typeof actionTypes.resetState;
}

export interface SetMethodAction extends Action {
  type: typeof actionTypes.setMethod;
  payload: string;
}
