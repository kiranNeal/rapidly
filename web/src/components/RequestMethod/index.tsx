import React, { useContext, useEffect } from 'react';
import { Dropdown } from 'carbon-components-react';

import './requestMethod.scss';
import { AppContext } from '../../App';
import { methodTypes } from '../../states/application/applicationTypes';
import * as applicationActions from '../../states/application/applicationActions';

const getMethodOption = (): Array<string> => {
  return Object.keys(methodTypes);
};

interface RequestMethodProps {
  initialMethod?: string;
}

const RequestMethod: React.FC<RequestMethodProps> = (
  props: RequestMethodProps
) => {
  const { state, dispatch } = useContext(AppContext);
  useEffect(() => {
    if (props.initialMethod) {
      dispatch(applicationActions.setMethod(props.initialMethod));
    }
  }, [dispatch, props.initialMethod]);
  return (
    <Dropdown
      id="request_method__dropdown"
      data-testid="request_method__dropdown"
      ariaLabel="Select Request Method"
      items={getMethodOption()}
      itemToString={(item: string): string => (item ? methodTypes[item] : '')}
      label=""
      titleText=""
      selectedItem={state.method}
      onChange={({ selectedItem }): void => {
        if (selectedItem) {
          dispatch(applicationActions.setMethod(selectedItem));
        }
      }}
    />
  );
};

export default RequestMethod;
