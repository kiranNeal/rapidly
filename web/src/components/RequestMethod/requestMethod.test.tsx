import React, { useReducer } from 'react';
import {
  render,
  cleanup,
  fireEvent,
  screen,
  getByLabelText
} from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';

import RequestMethod from './index';
import { AppContext } from '../../App';
import reducer, {
  initialState
} from '../../states/application/applicationReducer';
import * as applicationActions from '../../states/application/applicationActions';

interface WrapperProps {
  children: React.ReactNode;
}

const Wrapper: React.FunctionComponent<WrapperProps> = (
  props: WrapperProps
) => {
  const [state, dispatch] = useReducer(reducer, initialState);
  return (
    <AppContext.Provider value={{ state, dispatch }}>
      {props.children}
    </AppContext.Provider>
  );
};

describe('Request Method component tests', () => {
  afterEach(() => {
    cleanup();
  });

  it('Request method renders correctly with no initial Method prop passes', () => {
    const { asFragment } = render(
      <Wrapper>
        <RequestMethod />
      </Wrapper>
    );
    const container = document.body;
    expect(
      getByLabelText(container, 'Select Request Method')
    ).toHaveTextContent('GET');
    expect(asFragment()).toMatchSnapshot();
  });

  it('Request method renders correctly with initial Method and changes on select', async () => {
    render(
      <Wrapper>
        <RequestMethod initialMethod="post" />
      </Wrapper>
    );
    const container = document.body;
    expect(
      getByLabelText(container, 'Select Request Method')
    ).toHaveTextContent('POST');

    fireEvent.click(screen.getByText('POST'));
    const deleteOption = await screen.getByText('DELETE');
    fireEvent.click(deleteOption);

    expect(
      getByLabelText(container, 'Select Request Method')
    ).toHaveTextContent('DELETE');
  });
});
