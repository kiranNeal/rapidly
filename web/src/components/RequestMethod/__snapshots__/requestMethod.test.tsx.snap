// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Request Method component tests Request method renders correctly with no initial Method prop passes 1`] = `
<DocumentFragment>
  <div
    class="bx--dropdown__wrapper bx--list-box__wrapper"
  >
    <div
      aria-label="Select Request Method"
      class="bx--dropdown bx--list-box"
      id="request_method__dropdown"
      role="listbox"
      tabindex="-1"
    >
      <div
        aria-disabled="false"
        aria-expanded="false"
        aria-haspopup="true"
        aria-label="Open menu"
        class="bx--list-box__field"
        data-toggle="true"
        role="button"
        tabindex="0"
        type="button"
      >
        <span
          class="bx--list-box__label"
          for="downshift-0-input"
        >
          GET
        </span>
        <div
          class="bx--list-box__menu-icon"
        >
          <svg
            aria-label="Open menu"
            focusable="false"
            height="16"
            name="chevron--down"
            preserveAspectRatio="xMidYMid meet"
            role="img"
            style="will-change: transform;"
            viewBox="0 0 16 16"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M8 11L3 6 3.7 5.3 8 9.6 12.3 5.3 13 6z"
            />
            <title>
              Open menu
            </title>
          </svg>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
